_posts/2020-12-30-fast-jdbc-revisited.md

```diff
@@ -0,0 +1,252 @@
+                <artifactId>log4j-over-slf4j</artifactId>
+              </exclusion>
+        </exclusions>
+        </dependency>
+
+    </dependencies>
+    
+     <build>
+      <plugins>
+        <plugin>
+          <artifactId>maven-assembly-plugin</artifactId>
+          <configuration>
+            <archive>
+              <manifest>
+                <mainClass>com.uwekorn.Main</mainClass>
+              </manifest>
+            </archive>
+            <descriptorRefs>
+              <descriptorRef>jar-with-dependencies</descriptorRef>
+            </descriptorRefs>
+          </configuration>
+        </plugin>
+          <plugin>
+              <groupId>org.apache.maven.plugins</groupId>
+              <artifactId>maven-compiler-plugin</artifactId>
+              <version>3.8.0</version>
+              <configuration>
+                  <source>11</source>
+                  <target>11</target>
+              </configuration>
+          </plugin>
+      </plugins>
+    </build>
+</project>
+```
+
+We then start the JVM as before using the following code.
+
+```python
+import jpype
+import os
+
+classpath = os.path.join(os.getcwd(), "all-jar/target/drill-odbc-0.2-SNAPSHOT-jar-with-dependencies.jar")
+jpype.startJVM(jpype.getDefaultJVMPath(), f"-Djava.class.path={classpath}")
+```
+
+Before we can connect to the database, we must first start it using `./bin/drill-embedded`.
+Then we can connect to it using the client libraries:
+
+```python
+import jaydebeapi
+import jpype.dbapi2
+
+try:
+    # May fail on first attempt
+    jdbapi2_conn = jpype.dbapi2.connect("jdbc:drill:drillbit=127.0.0.1")
+except:
+    jdbapi2_conn = jpype.dbapi2.connect("jdbc:drill:drillbit=127.0.0.1")
+jdbapi2_cursor = jdbapi2_conn.cursor()
+
+try:
+    # May fail on first attempt
+    conn = jaydebeapi.connect('org.apache.drill.jdbc.Driver', 'jdbc:drill:drillbit=127.0.0.1')
+except:
+    conn = jaydebeapi.connect('org.apache.drill.jdbc.Driver', 'jdbc:drill:drillbit=127.0.0.1')
+cursor = conn.cursor()
+```
+
+With the newest combination of Drill and `jpype1`, you sadly get the following error when establishing the connection the first.
+On every second call, it passes though without any issues.
+
+```
+Exception: Java Exception
+
+The above exception was the direct cause of the following exception:
+
+java.lang.NoClassDefFoundError            Traceback (most recent call last)
+<ipython-input-5-51b73f9fbbd7> in <module>
+----> 1 jdbapi2_conn = jpype.dbapi2.connect("jdbc:drill:drillbit=127.0.0.1")
+
+…/env/lib/python3.8/site-packages/jpype/dbapi2.py in connect(dsn, driver, driver_args, adapters, converters, getters, setters, **kwargs)
+    418     # User supplied nothing
+    419     elif driver_args is None:
+--> 420         connection = DM.getConnection(dsn)
+    421 
+    422     # Otherwise use the kwargs
+
+java.lang.NoClassDefFoundError: java.lang.NoClassDefFoundError: oadd/org/apache/drill/exec/store/StoragePluginRegistry
+```
```

Comment:
JDBC has some strange behaviors as they do stuff with reflection that causes loading problems.  I haven't seen this one yet though I suspect it would be resolved with a JClass statement before the connection is established.   I would recommend putting in a ticket so that this one doesn't get lost.

Reply:
Adding a JClass statement didn't help, I'll also open a reproducible issue.

Reply:
https://github.com/jpype-project/jpype/issues/913